### ISSUE TITLE
semicolon_if_nothing_returned false positive inside unsafe block

### ISSUE THREAD

**Issue by dtolnay** [rust-lang/rust-clippy#7324]
In the following code, I believe it is correct as written, and a semicolon inside the block is not warranted.

```rust
use std::mem::MaybeUninit;
use std::ptr;

fn main() {
    let mut s = MaybeUninit::<String>::uninit();
    let _d = || unsafe { ptr::drop_in_place(s.as_mut_ptr()) };
}
```

```console
$ cargo clippy -- -Dclippy::pedantic

error: consider adding a `;` to the last statement for consistent formatting
 --> src/main.rs:6:26
  |
6 |     let _d = || unsafe { ptr::drop_in_place(s.as_mut_ptr()) };
  |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: add a `;` here: `ptr::drop_in_place(s.as_mut_ptr());`
  |
  = note: `-D clippy::semicolon-if-nothing-returned` implied by `-D clippy::pedantic`
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#semicolon_if_nothing_returned
```

Clippy currently wants the following instead, which is unnecessary and more verbose as formatted by rustfmt.

```rust
let _d = || unsafe {
    ptr::drop_in_place(s.as_mut_ptr());
};
```

FYI @1c3t3a @mbartlett21 

### Meta

- `cargo clippy -V`: clippy 0.1.54 (c79419a 2021-06-04)
- `rustc -Vv`:
  ```
  rustc 1.54.0-nightly (c79419af0 2021-06-04)
  binary: rustc
  commit-hash: c79419af0721c614d050f09b95f076da09d37b0d
  commit-date: 2021-06-04
  host: x86_64-unknown-linux-gnu
  release: 1.54.0-nightly
  LLVM version: 12.0.1
  ```

**Comment by 1c3t3a** [comment]
Mh yeah that's right. I am thinking about a rule here, maybe the lint should not fire in closures / unsafe blocks with just one statement? Or if the whole block is written on one line.

**Comment by mbartlett21** [comment]
Agreed.

I also had code like this, which I just ignored it on:

```rs
fn do_something() {
    unsafe { some_unsafe_op() };

    unsafe { some_other_unsafe_fn() };
}
```

The rule should be that it doesn't fire in an unsafe block, in an expression, with exactly one statement.

**Comment by 1c3t3a** [comment]
Alright, I would like to fix this if that's fine with you @mbartlett21?

**Comment by mbartlett21** [comment]
That's fine :)

**Comment by Enyium** [comment]
This lint is still triggered for me in unsafe blocks with one statement. All but one case have a function call inside that's formatted by rustfmt to put each argument on a separate line (up to 2 indents in `unsafe {}`). The one other and most concise case that still triggers this lint is:

```rust
unsafe {
    bridge::VideoInfo_GetFps(self.video_info.as_ref(), &mut numerator, &mut denominator)
};
```

Is it that rustfmt formats this on 3 instead of 1 line because of the length of the call, and then clippy wants a semicolon inside?

I think it would be inconsistent if you're forced to relocate the semicolon just based on the current length of the command, which could change.
